src/token/factory.sol

```diff
@@ -4,13 +4,31 @@ pragma solidity ^0.8.18;
 import {RestrictedToken} from "./restricted.sol";
 import {Memberlist} from "./memberlist.sol";
 
-interface RestrictedTokenFactoryLike {
-    function newRestrictedToken(string calldata, string calldata, uint8) external returns (address);
+interface ImmutableCreate2Factory {
+    function safeCreate2(bytes32 salt, bytes calldata initCode) external payable returns (address deploymentAddress);
 }
 
-contract RestrictedTokenFactory {
-    function newRestrictedToken(string memory name, string memory symbol, uint8 decimals) public returns (address) {
-        RestrictedToken token = new RestrictedToken(name, symbol, decimals);
+interface TrancheTokenFactoryLike {
+    function newTrancheToken(uint64, bytes16, string calldata, string calldata, uint8) external returns (address);
+}
+
+contract TrancheTokenFactory {
+    function newTrancheToken(uint64 poolId, bytes16 trancheId, string memory name, string memory symbol, uint8 decimals)
+        public
+        returns (address)
+    {
+        // Salt is hash(poolId + trancheId), to deploy copies of the restricted token contract
+        // on multiple chains with the same address for the same tranche
+        bytes32 salt = keccak256(abi.encodePacked(poolId, trancheId));
```

Comment:
The inputs to the deterministic address are
- Salt
- Deployer address
- Bytecode
- Arguments

The deployer address here is key. If we deploy the factory address, then any tranche tokens deployed from that factory address will be deterministic. If someone else deploys a tranche token with the same bytecode, salt and arguments, but from their own address or factory, then the deployer address will not match and it will lead to different addresses.

Reply:
Nice 👌